torchmetrics/text/squad.py

```diff
@@ -0,0 +1,198 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Any, Callable, Dict, Optional, Tuple
+
+import torch
+from torch import Tensor
+
+from torchmetrics import Metric
+from torchmetrics.functional.text.squad import PREDS_TYPE, TARGETS_TYPE, _squad_compute, _squad_update
+
+
+class SQuAD(Metric):
+    """Calculate `SQuAD Metric`_.
+
+    Args:
+        compute_on_step:
+            Forward only calls ``update()`` and returns None if this is set to False. default: True
+        dist_sync_on_step:
+            Synchronize metric state across processes at each ``forward()``
+            before returning the value at the step.
+        process_group:
+            Specify the process group on which synchronization is called. default: None (which selects the entire world)
+        dist_sync_fn:
+            Callback that performs the allgather operation on the metric state. When `None`, DDP
+            will be used to perform the allgather.
+
+    Example:
+        >>> from torchmetrics.text.squad import SQuAD
+        >>> predictions = [{"prediction_text": "1976", "id": "56e10a3be3433e1400422b22"}]
+        >>> references = [{"answers": {"answer_start": [97], "text": ["1976"]}, "id": "56e10a3be3433e1400422b22"}]
+        >>> _sqaud = SQuAD()
+        >>> _sqaud(predictions, references)
+        {'exact_match': tensor(100.), 'f1': tensor(100.)}
+
+    References:
+        [1] SQuAD: 100,000+ Questions for Machine Comprehension of Text by Pranav Rajpurkar, Jian Zhang, Konstantin
+        Lopyrev, Percy Liang `SQuAD Metric`_ .
+    """
+
+    higher_is_better = True
+
+    def __init__(
+        self,
+        compute_on_step: bool = True,
+        dist_sync_on_step: bool = False,
+        process_group: Optional[Any] = None,
+        dist_sync_fn: Optional[Callable] = None,
+    ):
+        super().__init__(
+            compute_on_step=compute_on_step,
+            dist_sync_on_step=dist_sync_on_step,
+            process_group=process_group,
+            dist_sync_fn=dist_sync_fn,
+        )
+
+        self.add_state(name="f1_score", default=[], dist_reduce_fx=None)
+        self.add_state(name="exact_match", default=[], dist_reduce_fx=None)
+        self.add_state(name="total", default=[], dist_reduce_fx=None)
```

Comment:
@stancld I have updated the implementation now.
I think in terms of efficiency the only upside is less memory usage because you will be storing a single float or int instead of lists of floats and ints.

Reply:
Also use tensor accumulation when ever possible. Much more memory efficient.